scipy/special/_round.h

```diff
@@ -0,0 +1,64 @@
+/*
+ * Functions for adding two double precision numbers with rounding to
+ * infinity or rounding to negative infinity without using <fenv.h>.
+ *
+ * References
+ * ----------
+ * [1] Hida, Li, Bailey, "Library for Double-Double and Quad-Double
+ *     Arithmetic."
+ */
+#ifndef ROUND_H
+#define ROUND_H
+
+#include <numpy/npy_math.h>
+#include "_c99compat.h"
+
+
+double add_round_up(double a, double b)
+{
+    double s, e, v;
```

Comment:
This would very much break with `-O3` since (IIRC) `-O3` enables unsafe floating point optimizations. Tests I was using to see if it works were mentioned here: https://github.com/scipy/scipy/pull/6653#issuecomment-251879577; would be possible to put the tests in SciPy itself if we only run them when we detect that `fenv` is present. Shall I?